gradle/java.gradle

```diff
@@ -34,6 +34,17 @@ plugins.withId('java') {
         classifier = 'sources'
         from sourceSets.main.java
     }
+    artifacts { archives sourceJar }
+
+    task generateJavadoc(type: Javadoc) {
+        source = sourceSets.main.allJava
```

Comment:
I'd like to clarify one thing, is this generating the javadocs for all classes across all targets to package for publishing? We have focused only on the client javadocs so far, which makes me wonder if we really want to include for all classes.

Reply:
Yes this is generating javadoc for all source.  Maven central publishing _requires_ javadoc JARs for each JAR published: [Requirements](http://central.sonatype.org/pages/requirements.html#supply-javadoc-and-sources)
It does state that you can upload empty JARs if you need to, but I don't see any reason not to just upload the actual javadoc for these jars.

    If, for some reason (for example, license issue or it's a Scala project), you can not provide -sources.jar or -javadoc.jar , please make fake -sources.jar or -javadoc.jar with simple README inside to pass the checking

Reply:
The main reason I can think of is that we have focused on cleaning up the javadocs for the client API, but not for internal classes (e.g., segment store and controller). I'm worried that the quality of the javadocs for internal classes might not be great. What do you think of generating empty javadoc for all artifacts except the client one? I can create an issue for fixing javadocs elsewhere , and once we do it, then we can start publishing them.

Reply:
Yeah, I agree with you, we don't have proper javadocs for all the components. It is not advisable to publish them without amending them.

Reply:
We can override the default java doc task in all modules to create empty docs,  leaving the client module.

Reply:
I don't really see why it matters.. the source is published and these are really just for ide enhancements if you would ever view javadoc in your IDE.

Reply:
I don't really follow this argument. The JavaDocs are generated from the code comments. If we are not happy with the quality of the code comments that is fine and we should correct that. If we publish them or not, the fact still remains about the quality of the comments.

The simple fact is publishing to maven central **requires** publishing a javadoc and source artifact. These are mostly so IDEs can retrieve them for tooling as there is no way to browse them. If we want to publish to maven central, we need to follow the conventions they lay out and include these. I see no reason why this would be a problem.